src/memory_manager.rs

```diff
@@ -853,4 +1032,104 @@ mod test {
         memory_1.read(0, &mut buf);
         assert_eq!(buf, vec![2; 1000]);
     }
+
```

Comment:
Have you considered the following scenarios:

1. upgrading the memory manager from V1 to V2
2. freeing an already free memory
3. freeing a memory that was never used
4. freeing a memory, then growing it

This is a big change, and we need to have solid test coverage.